### ISSUE TITLE
preview functionality broken on windows in master

### ISSUE THREAD

**Issue by blaenk** [jaspervdj/hakyll#153]
TL;DR: The arguably more efficient code introduced by @simukis in 5f6035b8 unfortunately breaks Hakyll on Windows:
1. Sites now need to be compiled with `-threaded` to avoid foreign code from 5f6035b8 (indirectly included through fs-notify) from blocking all other threads. It's not a big deal, I think, but I wasn't aware of this and took me a while to figure out. More info [here](http://www.haskell.org/ghc/docs/latest/html/libraries/base/Control-Concurrent.html#g:5):
   
   > Different Haskell implementations have different characteristics with regard to which operations block all threads.
   > 
   > Using GHC without the -threaded option, all foreign calls will block all other Haskell threads in the system, although I/O operations will not. With the -threaded option, only foreign calls with the unsafe attribute will block all other threads.
   
   Without `-threaded` on Windows, the notify code blocks all other threads (notably the server itself). This was apparent/obvious when I realized that the server only managed to load up (and serve requests) when I spam-saved a file under watch. I imagine after the callback for each event was called, the server was afforded a time slice.
2. Although `-threaded` allows things to operate as usual again, as soon as a file is modified (e.g. saved) an error is output in the form of:
   
   > site.exe: provider\posts/some-post.markdown: openFile: permission denied (Permission denied)
   
   Looking around online I've found [reason](http://stackoverflow.com/questions/6081794/openfile-permission-denied-when-reading-and-writing-to-the-same-file) to [believe](http://www.haskell.org/pipermail/haskell-cafe/2011-December/097984.html) that this has to do with lingering, un-closed handles as in the following:
   
   ``` haskell
   resourceString :: Provider -> Identifier -> IO String
   resourceString p i = readFile $ resourceFilePath p i
   ```
   
   Maybe the way the new preview system works, the handles from the first run-through/compilation process are still lingering during the next pass. I honestly have no idea, but I have a _feeling_ that it's not caused by the fs-notify code. In ghci I setup some file notifications with `watchTree`, the same function used by the Hakyll code, and was able to `openFile`/`readFile` perfectly fine.

I realize Windows is probably not a pressing priority. I primarily use and develop this on Linux, but I originally chose Hakyll in part because I was pleasantly surprised that it worked perfectly fine (4.2.2.0) on Windows.

I was doing perfectly fine on 4.2.2.0 but I wanted to use a custom provider directory. In 4.2.2.0, a custom provider directory made the preview functionality enter an infinite loop. I was looking around for the cause to this when I decided to try the master branch, where I learned the problem is fixed indirectly due to the introduction of @simukis' more efficient auto-compilation code in 5f6035b8. Everything worked perfectly fine on Linux.

I hope there's support for fixing this. I figure since everything worked perfectly fine on Windows in 4.2.2.0, there's no reason to regress. A crucial feature for me is to be able to work on my site while I'm on Windows.

I'll continue to look into the issue myself, but I'm not as experienced in debugging Haskell as you guys might be. I'd appreciate any ideas for leads or for debugging this. I already greatly miss the tools available to me on Linux (e.g. `strace`), since I'm not very familiar with debugging on Windows.

**How to reproduce**:

```
hakyll-init testing && cd testing
ghc --make -threaded site.hs
./site preview -v
# open about.rst with editor, save file (can edit if you want, but no difference in effect)
[ERROR] .\about.rst: openFile: permission denied (Permission denied)
```
